### DISCUSSION TITLE
When do I have to call `compinit` manually?

### DISCUSSION THREAD

**Discussion by borekb** [marlonrichert/zsh-snap#61] [Q&A] [answered]
[From README](https://github.com/marlonrichert/zsh-snap/tree/9786914fecf42948597abfe99436e26a3fc7997b#automatic-compinit-and-bashcompinit):

> **Automatic compinit and bashcompinit**
> You no longer need to put compinit or bashcompinit in your .zshrc file. Znap will run these for you as needed.

But I need to run it manually after e.g.

```
znap source zsh-users/zsh-completions
```

or

```
znap source agkozak/zsh-z
```

otherwise, completions don't work there.

I don't know much about how Zsh or completions work, could somebody explain what the README section really means and when I still need to run `compinit` manually?

---

**UPDATE**: there was a bug that is now fixed, see https://github.com/marlonrichert/zsh-snap/discussions/61#discussioncomment-560294.

**Comment by marlonrichert** [comment, marked as answer]
Short answer: _Never._ With Znap, you really _never_ have to call `compinit` manually. Any completions added to your `$fpath` while executing the code inside your `.zshrc` file will be automatically passed to `compinit`. 

Longer answer: Well, _theoretically_, you'd have to run `compinit` manually _if_ you would add completions to your `$fpath` _after_ your shell has started up. But there is no practical reason to ever do that. As long as you add your completions to your `$fpath` _inside your `.zshrc` file only,_ then there really is no reason to ever bother running `compinit` manually when using Znap.

So, long story short, for your case, just put the following into your `.zshrc` file:
```zsh
znap source agkozak/zsh-z
fpath+=( ~[zsh-users/zsh-completions]/src )
```
I just tested it and the completions work fine that way. (If it doesn't for you, please open a bug report for it.)

**Reply by borekb** [reply]
I probably still don't understand because this `.zshrc` doesn't work for me (the completions don't work):

```bash
# Source Znap
source ~/.znap/zsh-snap/znap.zsh

# Load prompt as soon as possible
znap prompt sindresorhus/pure

# Add Oh-My-Zsh
znap source ohmyzsh/ohmyzsh

# Somewhere later, completions:
znap source zsh-users/zsh-completions
znap source agkozak/zsh-z
```

Is this order wrong?

**UPDATE**: When I remove Oh My Zsh:

```bash
# Source Znap
source ~/.znap/zsh-snap/znap.zsh

# Load prompt as soon as possible
znap prompt sindresorhus/pure

# Completions
znap source zsh-users/zsh-completions
znap source agkozak/zsh-z
```

I get semi-functional completions – they do something although it looks different from what I'm used to and e.g. `Up` and `Down` keys don't work:

![Screen Shot 2021-04-02 at 8 33 42](https://user-images.githubusercontent.com/101152/113388261-2fb59300-938e-11eb-8deb-6c3d9ab5fb3b.png)

**UPDATE 2**: Oh, the above might actually be the default behavior of plain Zsh?! I've always used Oh My Zsh from day 1 and when I add it back _and_ call `compinit` manually (I have to do it to make completions work):

```bash
# Source Znap
source ~/.znap/zsh-snap/znap.zsh

# Load prompt as soon as possible
znap prompt sindresorhus/pure

# Add Oh-My-Zsh
znap source ohmyzsh/ohmyzsh

# Somewhere later, completions:
znap source zsh-users/zsh-completions
znap source agkozak/zsh-z

# Manual compinit
autoload -U compinit && compinit
```

I get the experience I'm used to:

![Screen Shot 2021-04-02 at 8 41 15](https://user-images.githubusercontent.com/101152/113388823-342e7b80-938f-11eb-99d8-a2aca291a0d9.png)

**UPDATE 3**: Moving Oh My Zsh towards the bottom also works – it probably calls `compinit` internally:

```bash
# Source Znap
source ~/.znap/zsh-snap/znap.zsh

# Load prompt as soon as possible
znap prompt sindresorhus/pure

# Completions quite high up in `.zshrc`
znap source zsh-users/zsh-completions
znap source agkozak/zsh-z

# Oh-My-Zsh last (calls `compinit`?!)
znap source ohmyzsh/ohmyzsh
```

---

So what am I doing "wrong", if anything? Oh My Zsh somehow seems significant for completions working or not – I prefer the UX of Oh My Zsh + completions but that seems to require a manual call to `compinit` (or adding Oh My Zsh last which probably calls `compinit` automatically?). Should I at any point be adding things to `fpath` manually?

I'd just like to understand what's going on and what's the right way. Thanks for your help 🙏.

**Reply by marlonrichert** [reply]
You're right: Completions don't seem to work correctly when I add `znap source ohmyzsh/ohmyzsh` to my `.zshrc` file. I'll have to investigate why.

**Reply by marlonrichert** [reply]
@borekb It should be fixed now (2a42781f8f34af1b6f57f4842fabb92036e04ce6, 546496899d40f762dc4b1f719eb9f24d7fcc43ec). Please update your `.zshrc` file with the following:
```zsh
source ~/.znap/zsh-snap/znap.zsh
znap prompt sindresorhus/pure               # Show instant prompt.
znap source ohmyzsh/ohmyzsh 'lib/*'         # Load OMZ's libs only. You don't need the rest.
fpath+=( ~[zsh-users/zsh-completions]/src ) # Add extra completions to $fpath.
znap source agkozak/zsh-z                   # Init `z` command.
```

**Reply by borekb** [reply]
@marlonrichert Thanks! A couple of observations:

**1:** I indeed don't need to call `compinit` manually and both zsh-z and zsh-completions work 🎉.

**2:** I need to source the whole OMZ to get "better completions" (see "UPDATE" vs. "UPDATE 2" screenshots above):

```diff
source ~/.znap/zsh-snap/znap.zsh
znap prompt sindresorhus/pure
-znap source ohmyzsh/ohmyzsh 'lib/*'
+znap source ohmyzsh/ohmyzsh
fpath+=( ~[zsh-users/zsh-completions]/src )
znap source agkozak/zsh-z
```

**3:** I don't quite understand this line:

```bash
fpath+=( ~[zsh-users/zsh-completions]/src )
```

I've never needed to deal with `fpath` before so this is new to me. The above seems to download (pull) the `zsh-users/zsh-completions` repo, similarly to the `znap source` command, but are you saying that it's necessary to do this `fpath` thing instead of a simpler `source`? This seems to work for me as well:

```bash
znap source zsh-users/zsh-completions
```

---

Overall, this is a `.zshrc` snippet that works for me:

```bash
source ~/.znap/zsh-snap/znap.zsh
znap prompt sindresorhus/pure
znap source ohmyzsh/ohmyzsh
znap source zsh-users/zsh-completions
znap source agkozak/zsh-z
```

**Reply by marlonrichert** [reply]
> **2:** I need to source the whole OMZ to get "better completions" (see "UPDATE" vs. "UPDATE 2" screenshots above):
> 
> ```diff
> source ~/.znap/zsh-snap/znap.zsh
> znap prompt sindresorhus/pure
> -znap source ohmyzsh/ohmyzsh 'lib/*'
> +znap source ohmyzsh/ohmyzsh
> fpath+=( ~[zsh-users/zsh-completions]/src )
> znap source agkozak/zsh-z
> ```

Sorry, it appears that I pushed in something that was slightly broken. Please update and try again. Both should produce the same result now.
 
> **3:** I don't quite understand this line:
> 
> ```bash
> fpath+=( ~[zsh-users/zsh-completions]/src )
> ```
> 
> I've never needed to deal with `fpath` before so this is new to me. The above seems to download (pull) the `zsh-users/zsh-completions` repo, similarly to the `znap source` command, but are you saying that it's necessary to do this `fpath` thing instead of a simpler `source`? This seems to work for me as well:
> 
> ```bash
> znap source zsh-users/zsh-completions
> ```

The two statements are (in this case) equivalent, because [`zsh-completions.plugin.zsh`](https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions.plugin.zsh) does the same thing under the hood. Of course you can use `znap source zsh-users/zsh-completions`, if that's what you prefer; the end result is the same. I personally just prefer to eliminate all unnecessary steps in between. 🙂

> Overall, this is a `.zshrc` snippet that works for me:
> 
> ```bash
> source ~/.znap/zsh-snap/znap.zsh
> znap prompt sindresorhus/pure
> znap source ohmyzsh/ohmyzsh
> znap source zsh-users/zsh-completions
> znap source agkozak/zsh-z
> ```

Sure, that's fine. `znap source ohmyzsh/ohmyzsh 'lib/*'` is slightly faster than `znap source ohmyzsh/ohmyzsh`, because it eliminates all unnecessary steps, but you're free to use whichever one you prefer.

**Comment by borekb** [comment]
_A question not to bury it in the [sibling thread](https://github.com/marlonrichert/zsh-snap/discussions/61#discussioncomment-560294):_

Znap seems to overwrite `compinit` to make it basically an empty function, so it seems to have zero cost compared to usual. I wonder if it's possibly a good idea to add this as the last line of my `.zshrc`:

```bash
# Last line of my `.zshrc:
autoload -U compinit && compinit
```

Would this make completions any safer? Like, if there are "bugs" in Znap (like [before fixes](https://github.com/marlonrichert/zsh-snap/discussions/61#discussioncomment-563648)) `compinit` would be called manually; and if Znap handles it via its magic, it's zero cost so there's no harm calling it?! At least that's what I'm thinking...

**Reply by marlonrichert** [reply]
* Calling `autoload compinit` after sourcing Znap won't do anything because `compinit()` is already defined. `autoload` will only load functions that haven't been defined yet.
* Calling `compinit` after sourcing Znap won't do anything, because (as you noticed) it's an empty function. Sure, there's no _harm_ in calling it, but it’s also completely useless because _it doesn't do anything._
* Calling the actual `compinit` can take _over a second._ You don't want to be doing that more than once during startup. That's why Znap prevents it from being called again.
* There is no need to run `compinit` more than once. After running `compinit`, you can always add additional functions to the completion system  by calling `compdef`.

**Reply by borekb** [reply]
Thank you, that explains it nicely.
